### PR TITLE
CMake: fix issue with version info not getting baked into LlamaConfig.cmake

### PR DESCRIPTION
**Pull request by mtmcp**
This change fixes an issue with the relocatable CMake package version info not being updated properly by separating the build-info query from generation of corresponding `common/build-info.cpp`.

### CODE REVIEW

**Comment by cebtenzzre** [comment]
Ah, so that's why we were generating build-info.cpp at configure time. I removed it because I thought it was redundant, but I didn't know about LlamaConfig.cmake.

File: scripts/build-info.cmake [hunk 1]
```diff
@@ -1,5 +1,4 @@
-set(TEMPLATE_FILE "${CMAKE_CURRENT_SOURCE_DIR}/common/build-info.cpp.in")
-set(OUTPUT_FILE "${CMAKE_CURRENT_SOURCE_DIR}/common/build-info.cpp")
+
```
Comment (cebtenzzre) [inline scripts/build-info.cmake:1]: remove this blank line

**Comment by bandoti** [comment]
> Ah, so that's why we were generating build-info.cpp at configure time. I removed it because I thought it was redundant, but I didn't know about LlamaConfig.cmake.

No problem! The test application doesn’t build because it is supposed to simulate an app outside the build tree. I want to (at some point) add a CI build for that to catch any regression issues in the future.

**Comment by cebtenzzre** [comment]
LLAMA_SHARED_LIB is blank in build/LlamaConfig.cmake when building shared libs is not enabled. Shouldn't BUILD_SHARED_LIBS be an `option` so it defaults to OFF?

https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html
> This variable is often added to projects as an [option()](https://cmake.org/cmake/help/latest/command/option.html#command:option) so that each user of a project can decide if they want to build the project using shared or static libraries.

**Comment by bandoti** [comment]
That's not what I'm seeing, which is explicitly set to `OFF` (using the changes in this commit). They should _not_ be options because `LlamaConfig.cmake` simply reflects how the user built llama.cpp. It is determined by options set in the top-level `CMakeLists.txt` file.
```
set(LLAMA_VERSION 0.0.1493)
set(LLAMA_BUILD_COMMIT 03223d6)
set(LLAMA_BUILD_NUMBER 1493)
set(LLAMA_SHARED_LIB OFF)
set(LLAMA_BLAS OFF)
set(LLAMA_CUBLAS OFF)
set(LLAMA_METAL OFF)
set(LLAMA_MPI OFF)
set(LLAMA_CLBLAST ON)
set(LLAMA_HIPBLAS OFF)
set(LLAMA_ACCELERATE ON)
```

**Comment by cebtenzzre** [comment]
I mean that it should be an option in the top-level CMakeLists.txt. I don't see where BUILD_SHARED_LIBS gets set to OFF unless it is done explicitly via the command line (and from then on, it will be cached until you rm -rf the build directory).

**Comment by bandoti** [comment]
> I mean that it should be an option in the top-level CMakeLists.txt. I don't see where BUILD_SHARED_LIBS gets set to OFF unless it is done explicitly via the command line (and from then on, it will be cached until you rm -rf the build directory).

Ah—gotcha. Yeah, I am passing it in on the command line: `-DBUILD_SHARED_LIBS=OFF`. I thought it was a generally supported option so didn't consider that.

**Comment by cebtenzzre** [comment]
I don't have push access but I suggest applying this patch:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 31908fe5..3e000941 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -43,6 +43,7 @@ else()
 endif()
 
 # general
+option(BUILD_SHARED_LIBS                "build shared libraries"                                OFF)
 option(LLAMA_STATIC                     "llama: static link libraries"                          OFF)
 option(LLAMA_NATIVE                     "llama: enable -march=native flag"                      ON)
 option(LLAMA_LTO                        "llama: enable link time optimization"                  OFF)
```

Otherwise, it seems fine.

**Review by cebtenzzre — APPROVED** [review]

**Pull request merged**